presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,129 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.cost.CostComparator;
+import com.facebook.presto.sql.analyzer.FeaturesConfig;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.Set;
+
+import static com.facebook.presto.SystemSessionProperties.getJoinDistributionType;
+import static com.facebook.presto.sql.planner.iterative.rule.ReorderJoinsUtils.createJoinAccordingToPartitioning;
+import static com.facebook.presto.sql.planner.iterative.rule.ReorderJoinsUtils.generatePartitions;
+import static com.facebook.presto.sql.planner.iterative.rule.ReorderJoinsUtils.planPriorityQueue;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.PARTITIONED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.REPLICATED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.google.common.base.Preconditions.checkState;
+
+public class ReorderJoins
+        implements Rule
+{
+    private final CostComparator costComparator;
+
+    public ReorderJoins(CostComparator costComparator)
+    {
+        this.costComparator = costComparator;
+    }
+
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        if (!(node instanceof JoinGraphNode)) {
+            return Optional.empty();
+        }
+
+        JoinGraphNode joinGraph = (JoinGraphNode) node;
+        JoinNode joinNode = new JoinEnumerator().chooseJoinOrder(joinGraph, idAllocator, symbolAllocator, lookup, session);
+        return Optional.of(joinNode);
+    }
+
+    private class JoinEnumerator
+    {
+        private final Map<JoinGraphNode, JoinNode> memo = new HashMap<>();
+
+        private JoinNode chooseJoinOrder(JoinGraphNode joinGraph, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Lookup lookup, Session session)
+        {
+            if (memo.containsKey(joinGraph)) {
+                return memo.get(joinGraph);
```

Comment:
I have a similar issue for not generating JoinGraphNodes when we've already done JoinReordering on the same tree (which right now has a hack to check if distribution type is set).    I asked @martin about it.  I'll see if the solution for that could apply for this too.  Otherwise I'll introduce an equivalence.